### PR TITLE
curl: add v8.4.0, allow r@8.3: to use it

### PR DESCRIPTION
**Pull request by haampie**

### CODE REVIEW

**Review by alecbcs — APPROVED** [review]
Confirmed SHA. Looks good to me. @haampie thanks for the quick update in response to the CVE.

File: var/spack/repos/builtin/packages/r/package.py [hunk 1]
```diff
@@ -70,9 +70,9 @@ class R(AutotoolsPackage):
     depends_on("blas", when="+external-lapack")
     depends_on("lapack", when="+external-lapack")
     depends_on("bzip2")
-    # R didn't anticipate the celebratory
-    # non-breaking major version bump of curl 8.
-    depends_on("curl+libidn2@:7")
+    depends_on("curl+libidn2")
+    # R didn't anticipate the celebratory non-breaking major version bump of curl 8.
+    depends_on("curl@:7", when="@:4.2")
```
Comment (alecbcs) [inline var/spack/repos/builtin/packages/r/package.py:75]: Does `r@4.2:` no longer depend on curl? Or is this a typo?

Comment (haampie) [reply var/spack/repos/builtin/packages/r/package.py:75]: It does, see line 73. Line 75 is just a forward compat bound

**Review by bernhardkaindl — APPROVED** [review]
@alecbcs You overlooked the unconditional
```py
    depends_on("curl+libidn2")
```
Check:
```py
bin/spack spec r
-   r@4.3.0%gcc@11.1.0~X~external-lapack~memory_profiling~rmath ...
...
-       ^curl@8.4.0%gcc@11.1.0~gssapi~ldap+libidn2~librtmp~libssh~libssh2+nghttp2 ...
```
And the
```py
    # R didn't anticipate the celebratory
    # non-breaking major version bump of curl 8.
    depends_on("curl@:7", when="@:4.2")
```
of course means that the additional restriction "curl@:7" is only needed
until r@4.2, but not above r@4.2.

Looks all good to me.

Passed: `bin/spack install -v r` built `r@4.3.0` using `curl@8.4.0` just fine on x86_64-focal.

**Pull request merged**